accepted/wasmtime-plugins.md

```diff
@@ -0,0 +1,328 @@
+
+In order to address the lack of WebAssembly adoption in third-party libraries, lacking WebAssembly features and allow integration with existing projects, I suggest yet another WASI proposal to be built into `wasmtime`: [wasi-dl].
+
+## [wasi-dl]
+
+[wasi-dl] is a WASI proposal, which ultimately abstracts over [`dlfcn.h`](https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/dlfcn.h.html) allowing components to link to arbitrary dynamic libraries at runtime and access symbols provided by them.
+
+### Interface
+
+Core functionality, extracted from https://github.com/rvolosatovs/wasi-dl/blob/6d2000d92d96b0967eb5a7ead314a765b7f596e2/wit/dl.wit looks as follows:
+
+```wit
+package wasi:dl@0.2.0-draft;
+
+interface ffi {
+    resource alloc {
+        new: static func(types: ffi-type) -> result<alloc>;
+
+        // Read and write accessors
+    }
+
+    enum primitive-type {
+        // Primitive C types
+    }
+
+    resource struct-type {
+        constructor(fields: list<ffi-type>);
+    }
+
+    variant ffi-type {
+        primitive(primitive-type),
+        struct(struct-type),
+    }
+
+    // FFI type definitions etc.
+}
+
+interface dll {
+    use ffi.{alloc, ffi-type, outgoing-value, incoming-value};
+
+    resource function {
+        /// Constructs a function from an opaque `alloc` and a type signature
+        /// Fails if type of `alloc` is not `ffi-type::primitive(primitive-type::pointer)`
+        from-alloc: static func(alloc: alloc, args: list<ffi-type>, ret: option<ffi-type>) -> result<function>;
+
+        /// Calls a function with specified arguments
+        /// Fails if argument types do not match function type signature
+        call: func(args: list<outgoing-value>) -> result<option<incoming-value>>;
+
+        /// Returns a function as an opaque `alloc` of `ffi-type::primitive(primitive-type::pointer)` type
+        get-alloc: func() -> alloc;
+    }
+
+    resource symbol {
+        /// Returns symbol as a typed `function`
+        get-function: func(args: list<ffi-type>, ret: option<ffi-type>) -> function;
+
+        /// Returns symbol as a typed `alloc`
+        get-alloc: func(ty: ffi-type) -> alloc;
+    }
+
+    /// `library::open` flags passed to `dlopen` by the host
+    flags open-flags {
+        lazy,
+        now,
+        global,
+        local,
+    }
+
+    resource library {
+        /// Opens a library using `dlopen`
+        open: static func(name: option<string>, flag: open-flags) -> result<library, string>;
+        // TODO: Add default() and next() ?
+
+        /// Gets a symbol from a library using `dlsym`
+        get: func(name: string) -> result<symbol, string>;
+    }
+
+    extension: func() -> string;
+    prefix: func() -> string;
+    suffix: func() -> string;
+}
+```
+
+Such interface is *unsafe* and it must be used with extreme care, however that is no different from any other host plugin, which would be loaded via `dlopen`. 
```

Comment:
To be more constructive, I would suggest an alternative approach that maintains a safe interface to Wasm, something like:

* There is some well-known symbol that plugin `.so`s should export, describing their WIT interface (maybe literally just a `static WIT_INTERFACE: &'static str = "..."` or alternatively the binary encoding of the same thing).
* Wasmtime loads a `plugin.so` and reads its WIT interface
* Wasmtime `dlsym`s the functions described by the WIT interface
* Wasmtime adds functions for that WIT interface to a `Linker`, these functions
  * translates Wasm / canonical ABI arguments into the equivalent in some sort of native ABI
  * call their corresponding `dlsym`ed functions from `plugin.so`
  * translate the native ABI's result back into Wasm / canonical ABI

In the above sketch, the `plugin.so` is trusted, but the Wasm is not. Any unsafety can only come from bugs in the `plugin.so` (either from its internal implementation or if its functions' types don't match the WIT interface it claims). Notably, unsafety cannot originate from within (untrusted and potentially malicious) Wasm guests, no matter what garbage values they indirectly pass to `plugin.so`.

The tricky parts here will be:

* What is the native ABI? Can we reuse the canonical ABI or a variant of it? I could imagine a `bindgen`-y proc macro that does some variant of the canonical ABI for plugins with statically-known interfaces, but what about dynamic interfaces (i.e. the common case for the `wasmtime` cli, rather than a `wasmtime` crate embedding that happens to use plugins of a certain shape)? What can we do to avoid arg/result translation overheads?
* A `plugin.so` may want some per-`Store` state, for example if `wasi-sockets` was implemented as a plugin, it would want any open sockets to be attached to the `Store`. How do we let `plugin.so` create that per-`Store` state? Where do we keep it? How do we pass it back to `plugin.so` on each call? How do we let `plugin.so` destroy it when we drop the store?
* Finally, it isn't clear to me whether this RFC proposes that `plugin.so`s are forwards compatible with new `wasmtime` versions (i.e. new Wasmtime releases are backwards compatible with old `plugin.so`s) or not. If so, then the ABI concerns described above are doubly important and we need to make sure they remain extensible for future additions and changes, which will involve a lot of subtleties.